### PR TITLE
Feature mes 3959 pass certificate validation improvements

### PR DESCRIPTION
**Pull request by rossfauldsbjss**
Added mod37 validation logic to ensure pass certificate is valid or display an error message if it isn't.

Note: due to tight time constraints I didn't create a provider but put the logic within the pass-finalisation page.  Due to significant differences between hotfix and develop I will refactor into a provider and associated unit tests.

### CODE REVIEW

File: src/pages/test-finalisation/pass-finalisation/pass-finalisation.ts [hunk 1]
```diff
@@ -231,26 +231,31 @@ export class PassFinalisationPage extends PracticeableBasePageComponent {
...
     return {
       provisionalLicenseProvidedCtrl: new FormControl(null, [Validators.required]),
       passCertificateNumberCtrl: new FormControl(null,
         {
-          validators: Validators.compose([Validators.maxLength(8), Validators.minLength(8), Validators.required]),
+          validators: Validators.compose([
+            Validators.maxLength(8),
+            Validators.minLength(8),
+            Validators.required,
+            this.validatePassCertificate.bind(this)]),
```
Comment (ammarhaiderbjss) [inline src/pages/test-finalisation/pass-finalisation/pass-finalisation.ts:251]: 🎉

**Review by edwardwilson — APPROVED** [review]

**Review by ammarhaiderbjss — APPROVED** [review]
Absolutely brilliant 🥇

File: src/pages/test-finalisation/pass-finalisation/pass-finalisation.ts [hunk 2]
```diff
@@ -260,6 +265,70 @@ export class PassFinalisationPage extends PracticeableBasePageComponent {
...
+    }
+
+    const checkDigit = this.calculateMod37CheckDigit(certificate);
+
+    if (checkDigit === 'invalid') {
+      return false;
+    }
+
+    // compare entered checkdigit with generated one
+    if (checkDigit !== certificate.toUpperCase()[7]) {
```
Comment (rstepnowski) [inline src/pages/test-finalisation/pass-finalisation/pass-finalisation.ts:293]: Are we certain that TARS won't care whether the checkDigit is in upper/lower case?

File: src/pages/test-finalisation/pass-finalisation/pass-finalisation.ts [hunk 3]
```diff
@@ -260,6 +265,70 @@ export class PassFinalisationPage extends PracticeableBasePageComponent {
...
+    }
+
+    return true;
+  }
+
+  isLetter(char: string): boolean {
+    return char.length === 1 && char.match(/[a-z]/i) !== null;
+  }
+
+  calculateMod37CheckDigit(certificate: string): string {
```
Comment (rstepnowski) [inline src/pages/test-finalisation/pass-finalisation/pass-finalisation.ts:304]: adjust the type-hinting since the function returns bool or a string

**Review by rstepnowski — APPROVED** [review]
Looks good 👍  - just some minor things ...

File: src/pages/test-finalisation/pass-finalisation/pass-finalisation.ts (hunk 2 above)
Comment (rossfauldsbjss) [reply src/pages/test-finalisation/pass-finalisation/pass-finalisation.ts:293]: we had a list of valid ones from submissions and it didn't care about case of check digit.

File: src/pages/test-finalisation/pass-finalisation/pass-finalisation.ts (hunk 3 above)
Comment (rossfauldsbjss) [reply src/pages/test-finalisation/pass-finalisation/pass-finalisation.ts:304]: it actually only returns string - either the checkDigit or the string 'invalid'.... no bool.

**Pull request merged**